### ISSUE TITLE
Backward through sparse_coo_tensor

### ISSUE THREAD

**Issue by ladzin** [pytorch/pytorch#49683]
## 🐛 Bug

Backward through torch.sparse_coo_tensor converts to a dense tensor

## To Reproduce

x = torch.ones((1), requires_grad=True)
S = torch.sparse_coo_tensor([[0],[0]], x, (100000, 100000))
torch.sparse.sum(S).backward() # calls to_dense(), allocating gigabytes of memory

## Expected behavior

Efficient backward using sparse tensors

## Environment

OS: Microsoft Windows 10 Enterprise
[pip3] numpy==1.19.2
[pip3] torch==1.8.0.dev20201219
[pip3] torchaudio==0.8.0.dev20201219
[pip3] torchvision==0.9.0.dev20201219
[conda] blas                      1.0                         mkl
[conda] cpuonly                   1.0                           0    pytorch-nightly
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38h196d8e1_0
[conda] mkl_fft                   1.2.0            py38h45dec08_0
[conda] mkl_random                1.1.1            py38h47e9c7a_0
[conda] numpy                     1.19.2           py38hadc3359_0
[conda] numpy-base                1.19.2           py38ha3acd2a_0
[conda] pytorch                   1.8.0.dev20201219     py3.8_cpu_0  [cpuonly]  pytorch-nightly
[conda] torchaudio                0.8.0.dev20201219            py38    pytorch-nightly
[conda] torchvision               0.9.0.dev20201219        py38_cpu  [cpuonly]  pytorch-nightly

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @aocsa @mruberry

**Comment by SsnL** [comment]
the gradient of sum is by definition dense, so a sparse representation would use even more memory. use `x.values().sum()` if you want sparse gradients.

**Comment by mruberry** [comment]
@pearu, how would we expect this to work after our latest discussions on gradients for sparse tensors?

**Comment by ngimel** [comment]
@ssnL, surprisingly, backward of `sum` is not a problem here, `Tensor.sparse.sum()` produces sparse gradient (unlike what mathematical definition of gradient for `sum` would require), with gradient only for original non-zero sparse elements specified. However, backward for `sparse_coo_tensor` quite inexplicably calls `to_dense` on this sparse gradient: 
https://github.com/pytorch/pytorch/blob/3779bdec56ca300ccb1e52efb76777ea574f1885/torch/csrc/autograd/FunctionsManual.cpp#L2777-L2786. 
Reliably fixing it would require enforcing `gradOutput indices are gradInput indices` invariant, as discussed in the sparse docs.

**Comment by SsnL** [comment]
Oh I see. I missed `torch.sparse.sum`. My bad.

On the `sparse_constructor_values_backward`, maybe I can explain it a bit since I wrote it. It should be more efficiently rewritten using a custom kernel for sparse tensor `index_select` (which need to use existing value if so and skip if not), but it did not exist when I wrote this part. If it still doesn't exist, a simple improvement can be done via creating a 0-1 sparse tensor and multiply it with the sparse gradient.

**Comment by ngimel** [comment]
As of last hour, kind of what you need exists here https://github.com/pytorch/pytorch/blob/12942ea52b6b36f34bf331e58c8d695b3be64bed/aten/src/ATen/native/sparse/cuda/SparseMatMul.cu#L767-L869, I don't know off hand if it's exactly what's needed. 
As for alternative, sparse-sparse multiplication is implemented, but it is ridiculously (to a tune of 100x slower than similarly-sized dense tensors) expensive. 
However, if we enforce invariant that gradients always have the same sparsity pattern as original tensors (which is current plan), then this selection/multiplication won't be needed in sparse_coo_tensor backward, as it will always receive a gradient with correct sparsity pattern which is can just pass through. This particular problem would move somewhere up the autograd chain :-)

**Comment by aocsa** [comment]
I would like to see this issue and open a PR.
I'll take a peek in the next day or two. cc @ngimel
